river/command/filter.zig

```diff
@@ -52,9 +49,82 @@ pub fn csdFilterAdd(
     args: []const []const u8,
     out: *?[]const u8,
 ) Error!void {
-    try appendFilter(
-        allocator,
-        &server.config.csd_filter,
-        args,
-    );
+    try modifyFilter(allocator, &server.config.csd_filter, args, .add);
+    csdFilterUpdateViews(args[1], .add);
+}
+
+pub fn csdFilterRemove(
+    allocator: *std.mem.Allocator,
+    seat: *Seat,
+    args: []const []const u8,
+    out: *?[]const u8,
+) Error!void {
+    try modifyFilter(allocator, &server.config.csd_filter, args, .remove);
+    csdFilterUpdateViews(args[1], .remove);
+}
+
+fn modifyFilter(
+    allocator: *std.mem.Allocator,
+    list: *std.ArrayList([]const u8),
+    args: []const []const u8,
+    operation: enum { add, remove },
+) Error!void {
+    if (args.len < 2) return Error.NotEnoughArguments;
+    if (args.len > 2) return Error.TooManyArguments;
+    for (list.items) |*filter, i| {
+        if (std.mem.eql(u8, filter.*, args[1])) {
+            if (operation == .remove) {
+                allocator.free(list.orderedRemove(i));
+            }
+            return;
+        }
+    }
+    if (operation == .add) {
+        try list.ensureUnusedCapacity(1);
+        list.appendAssumeCapacity(try std.mem.dupe(allocator, u8, args[1]));
+    }
+}
+
+fn csdFilterUpdateViews(app_id: []const u8, operation: enum { add, remove }) void {
+    // There is no link between Decoration and View, so we need to iterate over
+    // both separately. Note that we do not need to arrange the outputs here; If
+    // the clients decoration mode changes, it will receive a configure event.
```

Comment:
There is a link from Decoration to View, so there's no need to iterate over the views.
```zig
    const view = @ptrToInt(*View, decoration.xdg_toplevel_decoration.surface.data);
````